File: models/shopping_cart.cs
 namespace InternTask.models;
public class ShoppingCart
{
    private List<CartItem> items = new List<CartItem>();

    public void Add(Product product, int quantity)
    {
        if (product.AvailableQuantity < quantity)
            throw new InvalidOperationException($"Not enough quantity available for {product.Name}. Available: {product.AvailableQuantity}");
        
        items.Add(new CartItem(product, quantity));
    }

    public List<CartItem> GetItems()
    {
        return items;
    }

    public bool IsEmpty()
    {
        return !items.Any();
    }

    public decimal GetSubtotal()
    {
        return items.Sum(item => item.GetTotal());
    }

    public void Clear()
    {
        items.Clear();
    }
}

File: lib/ShoppingCart.java

class ShoppingCart {
    private java.util.List<CartItem> items = new java.util.ArrayList<>();

    public void add(Product product, int quantity) {
        if (product.getAvailableQuantity() < quantity)
            throw new IllegalStateException("Not enough quantity available for " + product.getName() + ". Available: " + product.getAvailableQuantity());
        
        items.add(new CartItem(product, quantity));
    }

    public java.util.List<CartItem> getItems() {
        return items;
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }

    public double getSubtotal() {
        double subtotal = 0;
        for (CartItem item : items) {
            subtotal += item.getTotal();
        }
        return subtotal;
    }

    public void clear() {
        items.clear();
    }
}